duckdb_engine/__init__.py

```diff
@@ -670,6 +672,47 @@ def _comment_query(  # type: ignore[no-untyped-def]
                 )
 
 
+def _serial_sequence_name(table_name: str) -> str:
+    return f"{table_name}_pk_sequence"
+
+
+@compiles(CreateTable, "duckdb")  # type: ignore[misc]
+def visit_create_table_with_serial_support(
+    instance: CreateTable, compiler: DDLCompiler, **kw: Any
+) -> str:
+    """Prepend a table with an auto-incremented primary key with the necessary sequence."""
+    table_sql = compiler.visit_create_table(instance, **kw)
+    table_name = instance.element.name
+    seq_name = _serial_sequence_name(table_name)
+    if seq_name not in table_sql:
+        # no need to create a sequence
+        return table_sql
+    seq_sql = f"\nCREATE SEQUENCE {seq_name};"
+    return seq_sql + table_sql
+
+
+@compiles(CreateColumn, "duckdb")  # type: ignore[misc]
+def visit_create_column_with_serial_support(
+    instance: CreateColumn, compiler: DDLCompiler, **kw: Any
+) -> str:
+    """Replace the first SERIAL field with a duckdb-style auto-incremented integer."""
+    column_sql = compiler.visit_create_column(instance, **kw)
+    if "SERIAL" not in column_sql or not kw.get("first_pk"):
+        return column_sql
+    table_name = instance.element.table.name
+    if "BIGSERIAL" in column_sql:
+        serial_type = "BIGSERIAL"
+        int_type = "BIGINT"
+    elif "SMALLSERIAL" in column_sql:
+        serial_type = "SMALLSERIAL"
+        int_type = "SMALLINT"
+    else:
+        serial_type = "SERIAL"
+        int_type = "INTEGER"
+    autoinc_type = f"{int_type} DEFAULT(nextval('{_serial_sequence_name(table_name)}'))"
+    return column_sql.replace(serial_type, autoinc_type)
```

Comment:
This is vulnerable to if there is the string `SERIAL` anywhere in the column_sql, eg if the column name is `MY_SERIAL_COLUMN`. Can you take a look to see if there is a way to inspect the `CreateColumn` instance and actually look for the datatype? If it's easy, I'd love to do this in a safer way. If it's not easy, then at least lets add a comment here disclosing this vulnerability.

Reply:
Similarly, above in visit_create_table, the line `if seq_name not in table_sql:` is vulnerable in the same way. If its easy to fix that vulnerability, that would be sweet.

Reply:
Yes, the detection via the name is clearly a weakness of this PR. I did not see another option, when I put this together, but I'll have another look.